content/blog/2019-11-19-Bril-phase-selection-and-ordering/index.md

```diff
@@ -0,0 +1,209 @@
+
+        We can simplify CFG by several transformations that eliminate useless edges and combine some basic blocks. Specifically, we can 1) replace a br instruciton with two identical destination to a jmp instruciton; 2) merging empty blocks; 3) merging two blocks with only one edge in between under some cases.
+
+    +   Tail Merging
+
+        If the predecesssor or successors share a same section of code in the end or begining correspondingly, we can change the jump to reuse the same segment of code.
+
+In general, the optimization we selected are tailored for the simple
+structure of Bril, which has a lot of contant calculation and branches that
+do not allow falling through. Passes like DCE and UCE will likely to remove
+more codes, but other passes might or might not generate new opportunities
+for DCE and UCE. Thus we need to order passes properly to obtain best
+performance, especially under the case where we can only apply limited number
+of passes.
+
+## Random Program Generator
+In order to analyze programs in Bril, one obstacle was that we didn't have a
+large enough corpus of code to analyze. To do statistical analysis or any
+kind of optimization techniques, we need a substantial training set to
+evaluate our ordering methods.
+
+To create this training set, we created a random Bril program generator. Our
+program generator evaluates each decision independently, and doesn't take any
+care to preserve program wide structure. We follow this flow chart at each
+step.
+
+1. First, we randomly chose whether we want to insert a binary operation, a constant, or an id.
+2. Next, given the instruction type, we decide whether it should be a boolean or an integer.
+3. Finally, we choose whether this instruction should write to a new identifier or an already existing identifier. If it's an already existing identifier we uniformly choose one from all the identifiers of the correct type.
+
+One thing that we've ignored so far is control flow. Considering that several
+of our optimizations are control flow optimizations, we clearly need to
+handle them somehow. However, introducing arbitrary control flow is fairly
+tricky. For example, we may construct non-terminating programs, which make
+evaluating them/checking our optimizations for bugs tricky.
+
+Thus, we settled on constructing a Directed Acyclic Graph. Every fixed
+interval, we insert a label. In addition to that label, we have a chance of
+adding a jump/branch to a succeeding label.
+
+We note that there are a lot of hand set probabilities in this generation
+process that have the ability to heavily affect the types of programs
+generated. For example, introducing a lot of control flow makes it more
+difficult for other optimizations to be applied. Thus, we select a mix of
+random constants.
+
+
+## Phase Ordering Analysis
+So, first of all, how much does the phase ordering even matter? To evaluate
+the effectiveness of our optimizations, we will use # of instructions as a
```

Comment:
Static instructions or dynamic instructions?

Reply:
Static. First of all, we don't have any loops, so static instruction count is morally equivalent to dynamic instruction count. Second, since all traces through our program are in some sense created equal, it becomes very unclear what inputs to use.

Clarified in post.